src/ProxyKit.Tests/RoundRobinTests.cs

```diff
@@ -48,5 +49,22 @@ public void Given_two_hosts_with_weight_then_round_robin_should_distribute_evenl
             upstreamHost = roundRobin.Next();
             upstreamHost.ToString().ShouldBe("http://localhost:5001/");
         }
+
+        [Fact]
+        public void Given_one_host_then_ensure_lock_is_released()
+        {
+            var roundRobin = new RoundRobin
+            {
+                new UpstreamHost("http://localhost:5001/")
+            };
+            var upstreamHost = roundRobin.Next();
+            upstreamHost.ToString().ShouldBe("http://localhost:5001/");
+            var thread = new Thread(() =>
+            {
+                roundRobin.Add(new UpstreamHost("http://localhost:5002/"));
+            });
+            thread.Start();
+            thread.Join(TimeSpan.FromSeconds(1)).ShouldBe(true);
```

Comment:
Under normal circumstances adding to the collection should definitely take less than a second here. If the thread waits too long on the lock for this case; then the lock probably hasn't been given back, so we can assume the lock is lost and the test should fail.